Add card component

buck06191 opened the pull request:
## What does this change?
<!-- A PR should have enough detail to be understandable far in the future. e.g what is the problem/why is the change needed, how does it solve it and any questions or points of discussion. Prefer copying information from a Trello card over linking to it; the card may not always exist and reviewers may not have access to the board. -->
Adds a `Card` component and some basic page layout structure.

`Card`s should act as the basic module for selecting a newsletter and providing information about it. The email sign up will be a separate component which should read the current selection of newsletters from the page. Cards are grouped by section, with accordion drawers holding each of these.

I've not yet added in the `User` functionality but have added some basic signposting around where I expect the state to be held, used and managed.

Remaining questions: 
- How do we do stories from components such as `Card`, where the state management happens in a higher level component?
- How are we going to handle signing up to multiple emails? Does the current data model for this work?
- What will the `User` information look like and how is this passed through?

### Screenshots

![image](https://user-images.githubusercontent.com/9122944/99975948-c0ee7a80-2d9a-11eb-94ff-a81ddc5c825b.png)


## How to test
<!-- Provide instructions to help others verify the change. This could take the form of "On PROD, do X and witness Y. On this branch, do X and witness Z. " -->
`yarn storybook` on a local branch is the best way to test it interactively

buck06191 committed:
Add card component

Add first basic component for newsletters round up page

buck06191 committed:
Add basic data types

buck06191 committed:
Implement styles

buck06191 committed:
Add src-layout for page structure;

buck06191 committed:
Add sign upa nd info components for cards

Add stories and components for the two separate chunks of a card. This
allows us to test these in isolation, which is important as we trial
different approaches including a single email field.

buck06191 committed:
Remove autogenerated test file until needed

buck06191 committed:
Add state management

State management is added to handle selection of newsletters. This
involved creating a few layers of wrappers and moving the email form
input out of the cards, to be replaced by a checkbox.

buck06191 committed:
Remove pre-push test check until tests added;

buck06191 committed:
Allow passing with no tests for now

buck06191 committed:
Add basic initial tests

alexduf commented on src/newsletters.ts:
```diff
@@ -0,0 +1,49 @@
+	frequency: 'Weekly',
+};
+
+export const newsletters: Newsletters = {
```
Am I correct that this is test data for now, awaiting further implementation?

buck06191 replied:
Correct. One of the key things I want to get working next would be setting this up with the Identity API in some format to pull this data through. If we can get this to be a truly server-side rendered project, then we can reduce the number of calls to the API right down so that it's cached in some format, though I haven't thought it through in too much detail yet.

The same will go for User info as well

alexduf replied:
Sounds good. The newsletter data could be injected on the server, but the user data would have to be fetched asynchronously on the client as it shouldn't be cached

alexduf reviewed (approved):
Looks great so far!

SiAdcock commented on src/App.tsx:
```diff
@@ -1,3 +1,18 @@
-import React from 'react';
+// this comment tells babel to convert jsx to calls to a function called jsx instead of React.createElement
```
I'm surprised you need this if you're using `@emotion/babel-preset-css-prop`. Is this needed for compilation, or for integration with other projects?

buck06191 replied:
Storybook wouldn't work without this. I've no idea why not, but it wasn't happy unless I added this in. I'd love to get it working without it because I've not seen any other Guardian projects need it

SiAdcock replied:
Getting Babel and Storybook to play together has been a bit weird. Storybook respects `.babelrc` but I'm not sure about `.babelrc.js`. You might need to create a [`.storybook/.babelrc`](https://storybook.js.org/docs/react/configure/babel#custom-config-file) file. For legacy reasons I can't quite remember, Source creates a [custom Webpack config](https://github.com/guardian/source/blob/da9b2181a2ce0e8a8b3eeccc3ddd716d1a31e0f3/.storybook/webpack.config.js#L11) 😱

buck06191 replied:
Swapping to `.babelrc` from `.babelrc.js` fixed this.

SiAdcock commented on src/components/Section.tsx:
```diff
@@ -0,0 +1,18 @@
+
+export const Section: React.FC<SectionProps> = ({ sectionName, children }) => (
+	<div data-testid="section">
+		<AccordionRow label={sectionName}>
```
[We recommend](http://theguardian.design/2a1e5182b/p/38c5aa-accordion/b/92b71e) wrapping your `<AccordionRow>` in an `<Accordion>`, if only to add the border to the bottom. Let me know if you have any questions about this 🙂

buck06191 replied:
@SiAdcock  I had that originally but had a problem with using `data-testid` with `AccordionRow` and couldn't find another way to count how many `Section` components appeared for the test I put in place. 

Do you know if it's possible to add `data-*` attributes to the Source components?

buck06191 replied:
Solution was to place the `<Section>` component inside `<AccordionRow>` and make it a `<div>` with the `data-testid` attribute around the `<Tiles>` component.

SiAdcock replied:
> Do you know if it's possible to add data-* attributes to the Source components?

Of course! For every component, except Accordion! 😅 

This is a bug, I'll get it fixed. Sorry for the inconvenience.

buck06191 committed:
Upgrade Source to fix bug

buck06191 committed:
Separate Section and Card out further

Separate out the Card and Section components that are used on Page such
that they function outside of the Page component. This allows for better
testing and better understanding of the code.

I've also moved the AccordionRow to being part of the Page component and
placed Section inside this to maintain testing whilst using the
component as the design system intends.

buck06191 committed:
Remove need for JSX pragma
